controllers/nodefeaturediscovery_status.go

```diff
@@ -0,0 +1,551 @@
+	// It may still be available, but it is degraded
+	ConditionDegraded string = "Degraded"
+
+	// ConditionUpgradeable indicates whether the resources maintained by the operator are in a state that is safe to upgrade.
+	// When `False`, the resources maintained by the operator should not be upgraded and the
+	// message field should contain a human readable description of what the administrator should do to
+	// allow the operator to successfully update the resources maintained by the operator.
+	ConditionUpgradeable string = "Upgradeable"
+)
+
+// updateStatus is used to update the status of a resource (e.g., degraded,
+// available, etc.)
+func (r *NodeFeatureDiscoveryReconciler) updateStatus(nfd *nfdv1.NodeFeatureDiscovery, condition []metav1.Condition) error {
+
+	// The actual 'nfd' object should *not* be modified when trying to
+	// check the object's status. This variable is a dummy variable used
+	// to set temporary conditions.
+	nfdCopy := nfd.DeepCopy()
+
+	// If a set of conditions exists, then it should be added to the
+	// 'nfd' Copy.
+	if condition != nil {
+		nfdCopy.Status.Conditions = condition
+	}
+
+	// Next step is to check if we need to update the status
+	modified := false
+
+	// Because there are only four possible conditions (degraded, available,
+	// updatable, and progressing), it isn't necessary to check if old
+	// conditions should be removed.
+	for _, newCondition := range nfdCopy.Status.Conditions {
+		oldCondition := meta.FindStatusCondition(nfd.Status.Conditions, newCondition.Type)
+		if oldCondition == nil {
+			modified = true
+			break
+		}
+		// Ignore timestamps to avoid infinite reconcile loops
+		if oldCondition.Status != newCondition.Status ||
+			oldCondition.Reason != newCondition.Reason ||
+			oldCondition.Message != newCondition.Message {
+			modified = true
+			break
+		}
+	}
+
+	// If nothing has been modified, then return nothing. Even if the list
+	// of 'conditions' is not empty, it should not be counted as an update
+	// if it was already counted as an update before.
+	if !modified {
+		return nil
+	}
+	return r.Status().Update(context.TODO(), nfdCopy)
+}
+
+// updateDegradedCondition is used to mark a given resource as "degraded" so that
+// the reconciler can take steps to rectify the situation.
+func (r *NodeFeatureDiscoveryReconciler) updateDegradedCondition(nfd *nfdv1.NodeFeatureDiscovery, condition string, conditionErr error) (ctrl.Result, error) {
+
+	// It is already assumed that the resource has been degraded, so the first
+	// step is to gather the correct list of conditions.
+	var conditionErrMsg string = "Degraded"
+	if conditionErr != nil {
+		conditionErrMsg = conditionErr.Error()
+	}
+	degradedCondition := r.getDegradedConditions(condition, conditionErrMsg)
+	if err := r.updateStatus(nfd, degradedCondition); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	return reconcile.Result{Requeue: true}, nil
+}
```

Comment:
Suggest to simplify

```suggestion
func (r *NodeFeatureDiscoveryReconciler) updateDegradedCondition(nfd *nfdv1.NodeFeatureDiscovery, reason, message string) (ctrl.Result, error) {
	degradedCondition := r.getDegradedConditions(reason, message)
	if err := r.updateStatus(nfd, degradedCondition); err != nil {
		return reconcile.Result{}, err
	}

	return reconcile.Result{Requeue: true}, nil
}
```

Ditto X times to all `updateXCondition` methods